src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala

```diff
@@ -20,91 +20,87 @@ import scala.collection.immutable.Set
-    }
-  }
-
-  override def build(containerType: ContainerInfo.Type, builder: CommandInfo.Builder, ev: ExternalVolume): Unit = {
-    // special behavior for docker vs. mesos containers
-    // - mesos containerizer: serialize volumes into envvar sets
-    if (containerType == ContainerInfo.Type.MESOS) {
-      val env = if (builder.hasEnvironment) builder.getEnvironment.toBuilder else Environment.newBuilder
-      val toAdd = volumeToEnv(ev, env.getVariablesList.asScala)
-      env.addAllVariables(toAdd.asJava)
-      builder.setEnvironment(env.build)
-    }
-  }
-
-  val driverOption = "dvdi/driver"
-  val quotedDriverOption = '"' + driverOption + '"'
-  val dvdiVolumeContainerPath = "DVDI_VOLUME_CONTAINERPATH"
-  val dvdiVolumeName = "DVDI_VOLUME_NAME"
-  val dvdiVolumeDriver = "DVDI_VOLUME_DRIVER"
-  val dvdiVolumeOpts = "DVDI_VOLUME_OPTS"
-
-  private[providers] def volumeToEnv(
-    vol: ExternalVolume,
-    i: Iterable[Environment.Variable]): Iterable[Environment.Variable] = {
-
-    import OptionLabelPatterns._
-
-    val suffix = {
-      val offset = i.filter(_.getName.startsWith(dvdiVolumeName)).map{ s =>
-        val ss = s.getName.substring(dvdiVolumeName.length)
-        if (ss.length > 0) ss.toInt else 0
-      }.foldLeft(-1)((z, i) => if (i > z) i else z)
-
-      if (offset >= 0) (offset + 1).toString else ""
-    }
-
-    def mkVar(name: String, value: String): Environment.Variable =
-      Environment.Variable.newBuilder.setName(name).setValue(value).build
-
-    val vars = Seq[Environment.Variable](
-      mkVar(dvdiVolumeContainerPath + suffix, vol.containerPath),
-      mkVar(dvdiVolumeName + suffix, vol.external.name),
-      mkVar(dvdiVolumeDriver + suffix, vol.external.options(driverOption))
-    )
-
-    val optsVar = {
+    def dockerVolumeParameters(volume: ExternalVolume): Seq[Parameter] = {
+      import OptionLabelPatterns._
       val prefix: String = name + OptionNamespaceSeparator
       // don't let the user override these
       val ignore = Set(driverOption)
       // external.size trumps any user-specified dvdi/size option
-      val opts = vol.external.options ++ Map[String, String](
-        vol.external.size.map(prefix + "size" -> _.toString).toList: _*
+      val opts = volume.external.options ++ Map[String, String](
+        volume.external.size.map(prefix + "size" -> _.toString).toList: _*
       )
 
       // forward all dvdi/* options to the dvdcli driver, stripping the dvdi/ prefix
       // and trimming the values
       opts.filterKeys{ k =>
         k.startsWith(prefix) && !ignore.contains(k.toLowerCase)
       }.map{
-        case (k, v) => k.substring(prefix.length) + "=" + v.trim()
-      }.mkString(",")
+        case (k, v) => Parameter.newBuilder
+          .setKey(k.substring(prefix.length))
+          .setValue(v.trim())
+          .build
+      }.toSeq
+    }
+
+    def applyOptions(dv: MesosDockerVolumeBuilder, opts: Seq[Parameter]): MesosDockerVolume = {
+      if (opts.isEmpty) {
```

Comment:
done

On Mon, May 2, 2016 at 8:52 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In
> src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala
> <https://github.com/mesosphere/marathon/pull/3832#discussion_r61734457>:
>
> >        )
> >
> >        // forward all dvdi/* options to the dvdcli driver, stripping the dvdi/ prefix
> >        // and trimming the values
> >        opts.filterKeys{ k =>
> >          k.startsWith(prefix) && !ignore.contains(k.toLowerCase)
> >        }.map{
> > -        case (k, v) => k.substring(prefix.length) + "=" + v.trim()
> > -      }.mkString(",")
> > +        case (k, v) => Parameter.newBuilder
> > +          .setKey(k.substring(prefix.length))
> > +          .setValue(v.trim())
> > +          .build
> > +      }.toSeq
> > +    }
> > +
> > +    def applyOptions(dv: MesosDockerVolumeBuilder, opts: Seq[Parameter]): MesosDockerVolume = {
> > +      if (opts.isEmpty) {
>
> A comment behind clearDriverOptions will help here.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mesosphere/marathon/pull/3832/files/a37d3f7eb4767cce78e17b6527de71b40d73d0a6#r61734457>
>